✨ Add ipv6 support to capd

mcwumbly opened the pull request:
quickstart e2e test can be run with:
```
IP_FAMILY=IPv6 \
DOCKER_SERVICE_CIDRS="fd00:100:64::/108" \
DOCKER_POD_CIDRS="fd00:100:96::/48" \
make test-e2e
```
Note: `POD_SUBNET` in kindnet.yaml is now set automatically based on `DOCKER_POD_CIDRS`

`IPFamily()` method on `Cluster` determines the ip family based on the service CIDRs and pod CIDRs. This is used by the docker provider to configure the docker machines accordingly.

Related feature request: https://github.com/kubernetes-sigs/cluster-api/issues/4557

cc @christianang

mcwumbly commented:
Note: Using serviceCIDRs to auto-detect whether a cluster is IPv4 or IPv6 is probably not the best approach since dual stack clusters _may_ only have one of these values set, per [KEP 563](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/563-dual-stack#services):

> A dual-stack cluster can have this flag configured as either `--services-cluster-ip-range=<cidr>,<cidr>` or `--services-cluster-ip-range=<cidr>`

In the underlying KEP, based on my current reading so far, it seems like it is non-trivial (and perhaps impossible) to detect the IP Family of a cluster's pod network at config-time.

Perhaps we have more control over that in Cluster API. For example, we could use `spec.ClusterNetwork.Pods`. But that field is [also optional](https://github.com/kubernetes-sigs/cluster-api/blob/402a4524f00671eabe6441c71d749dfe258bc43f/api/v1alpha4/cluster_types.go#L78-L80), so perhaps we can't depend on it.

Maybe all of this is arguing for either:

1. A new, explicit flag at the Cluster level or...
2. Leave it to the user to wire the right values through at the provider level

Will need to think about this, but interested in others' thoughts... If we go with (2), in the case of CAPD, where might the best place be for that configuration?

fabriziopandini commented on kind-up.sh:
```diff
@@ -0,0 +1,83 @@
+#!/bin/sh
```
If we are going to add this utility we should get this somewhere else (might be under scripts or under test) and document it. Given that this is not core of the goal of this PR I suggest to remove it from this changes set and eventually to address it in a separated PR

mcwumbly replied:
I'll plan to remove this from this PR...

fabriziopandini commented on test/e2e/config/docker.yaml:
```diff
@@ -92,9 +92,9 @@ variables:
   KUBERNETES_VERSION_UPGRADE_FROM: "v1.18.2"
   DOCKER_SERVICE_DOMAIN: "cluster.local"
-  DOCKER_SERVICE_CIDRS: "10.128.0.0/12"
+  DOCKER_SERVICE_CIDRS: "fd00:100:64::/108"
```
This file is being used by all the tests, so I think we should avoid to change the CIDR here; instead we can use one of the two alternatives:
- overriding the default value using env variables provided before running the test (all the test will run in ipv6 mode)
- overriding the default value using env variables set inside a specific test (only a specific test will run in ipv6 mode)

fabriziopandini replied:
resolving because I have summarised consideration about ipv6 E2E testing down in a comment

fabriziopandini commented on test/e2e/data/cni/kindnet/kindnet.yaml:
```diff
@@ -78,7 +78,7 @@ spec:
                   fieldPath: status.podIP
             - name: POD_SUBNET
-              value: "192.168.0.0/16"
+              value: "fd00:100:96::/48"
```
I think we should use variable substitution here, so kindnet manifest will adapt to what is being used for the cluster templates.
However I need to check if the machinery for CNI plugins support this and eventually make this possible (it is following a different code path due to some problems found in Azure due to a super long CNI yaml)

fabriziopandini replied:
resolving because I have summarised consideration about ipv6 E2E testing down in a comment

fabriziopandini commented on test/e2e/data/infrastructure-docker/v1alpha4/bases/cluster-with-kcp.yaml:
```diff
@@ -62,14 +62,21 @@ spec:
+        certSANs: [localhost, 127.0.0.1, 0.0.0.0, "::", "::1", host.docker.internal]
     initConfiguration:
+      localAPIEndpoint:
+        advertiseAddress: '::'
```
Assuming that we will run tests both in ipv4 and ipv6 modes, this change could have some tricky impacts...

We should think if there is a way to avoid us to duplicate templates, and if this is not possible and more variants of the templates will exists, we should think how to make the tests aware of which test to use...

fabriziopandini replied:
resolving because I have summarised consideration about ipv6 E2E testing down in a comment

fabriziopandini commented on test/e2e/data/infrastructure-docker/v1alpha4/bases/cluster-with-kcp.yaml:
```diff
@@ -62,14 +62,21 @@ spec:
-        kubeletExtraArgs: {eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'}
+        kubeletExtraArgs:
+          eviction-hard: 'nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%'
+          node-ip: "::"
```
Same as above

fabriziopandini replied:
resolving because I have summarised consideration about ipv6 E2E testing down in a comment

fabriziopandini commented on test/framework/bootstrap/kind_provider.go:
```diff
@@ -121,6 +121,9 @@ func withDockerSockConfig() *kindv1.Cluster {
 			Kind:       "Cluster",
 		},
+		Networking: kindv1.Networking{
+			IPFamily: kindv1.IPv6Family,
```
We should define how to make this info flow down the chain.
Also, are you expecting to support combination where the management cluster in kind has a different  ipFamily than the workload clusters or are we assuming that both are always consistent?

mcwumbly replied:
> are you expecting to support combination where the management cluster in kind has a different ipFamily than the workload clusters or are we assuming that both are always consistent?

I think in the short term, we should expect them to be consistent. Longer term, when we're really thinking about dual stack, I think you argue that the management cluster IP family should "contain" the workload cluster IP family.

fabriziopandini commented on test/infrastructure/docker/controllers/dockercluster_controller.go:
```diff
@@ -70,9 +70,14 @@ func (r *DockerClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	// Create a helper for managing a docker container hosting the loadbalancer.
-	externalLoadBalancer, err := docker.NewLoadBalancer(cluster.Name)
+	externalLoadBalancer, err := docker.NewLoadBalancer(cluster.Name, ipv6)
```
nit: might be worth to consider passing the entire cluster object to `docker.NewLoadBalancer` instead of passing two info derived from cluster

fabriziopandini commented on test/infrastructure/docker/controllers/dockermachine_controller.go:
```diff
@@ -137,8 +137,15 @@ func (r *DockerMachineReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+
 	// Create a helper for managing the docker container hosting the machine.
-	externalMachine, err := docker.NewMachine(cluster.Name, machine.Name, dockerMachine.Spec.CustomImage, nil)
+	externalMachine, err := docker.NewMachine(cluster.Name, machine.Name, dockerMachine.Spec.CustomImage, nil, ipv6)
```
nit same as above, consider if passing the entire cluster object

fabriziopandini commented on test/infrastructure/docker/docker/kind_manager.go:
```diff
@@ -124,6 +124,8 @@ func createNode(name, image, clusterLabel, role string, mounts []v1alpha4.Mount,
 		"--label", clusterLabel,
 		// label the node with the role ID
 		"--label", fmt.Sprintf("%s=%s", nodeRoleLabelKey, role),
+		"--sysctl=net.ipv6.conf.all.disable_ipv6=0",
+		"--sysctl=net.ipv6.conf.all.forwarding=1",
```
We should check kind default here or make this dynamic according to the cluster ipFamily

mcwumbly replied:
related code in kind: https://github.com/kubernetes-sigs/kind/blob/b6bc112522651d98c81823df56b7afa511459a3b/pkg/cluster/internal/providers/docker/provision.go#L192-L195

fabriziopandini replied:
Thanks for the pointer! This clarifies this bit to me

fabriziopandini commented on test/infrastructure/docker/docker/loadbalancer.go:
```diff
@@ -59,6 +61,7 @@ func NewLoadBalancer(name string) (*LoadBalancer, error) {
 	return &LoadBalancer{
 		name:      name,
 		container: container,
+		ipv6:      ipv6,
```
Might be work to keep IpFamily here (it depends by what is the expected behaviour for dual mode clusters)

fabriziopandini reviewed (commented):
@mcwumbly thanks for this PR!
I'm super happy to see the required changes for supporting ipv6 in CAPI are minimal.
Apart few nits, most of the questions are about what are the expectations for the target test framework: test only ipv6 or test ipv4 & ipv6 (and dual in future). Support mixed tests within the same job or not, etc.
Happy to discuss details if you find this useful

mcwumbly commented on api/v1alpha4/cluster_types.go:
```diff
@@ -215,6 +216,43 @@ func (c *Cluster) SetConditions(conditions Conditions) {
 	c.Status.Conditions = conditions
 }
 
+func (c *Cluster) IPFamily() (ClusterIPFamily, error) {
```
I see... if we want to continue to pursue the auto-detection route, then I think this may benefit from a table test of sorts:

When service CIDR and pod CIDR are both specified and match there is no ambiguity.

I think these cases are also pretty safe to determine:

| service CIDR | pod CIDR | cluster IP Family | note |
| -- | -- | -- | -- |
| (empty) | ipv4 | ipv4 |
| (empty) | ipv6 | ipv6 |
| (empty) | ipv4,ipv6 | dualstack |
| ipv4,ipv6 | (empty) | dualstack |
| ipv4 | ipv4,ipv6 | dualstack |
| ipv6 | ipv4,ipv6 | dualstack |


When they are both specified an IP Family within service CIDR list is not in the pod CIDR list, then it should be invalid, for example:

| service CIDR | pod CIDR | cluster IP Family | note |
| -- | -- | -- | -- |
| ipv4 | ipv6 | invalid | 
| ipv4,ipv6 | ipv6 | invalid | 
| ipv6 | ipv4 | invalid | 
| ipv4,ipv6 | ipv6 | invalid | 

It is just these cases that I believe are ambiguous, as they could be dualstack. Perhaps in these cases we could choose a default as shown?

| service CIDR | pod CIDR | cluster IP Family | note |
| -- | -- | -- | -- |
| (empty) | (empty) | ipv4 | could be dualstack
| ipv4 | (empty) | ipv4 | could be dualstack
| ipv6 | (empty) | ipv6 | could be dualstack

CecileRobertMichon replied:
This is what we have in CAPZ currently to support single stack IPv6 on Azure: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/templates/cluster-template-ipv6.yaml

this is the workaround we had to apply to get it working: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/templates/cluster-template-ipv6.yaml#L147-L150

+1 to auto-detection, it'd be great if we can avoid requiring changes for providers

CecileRobertMichon replied:
Also what we've done so far is check the vnet CIDR to determine if it's using ipv6: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/azure/scope/cluster.go#L301

mcwumbly replied:
Oh, thanks for pointing that out. I see there's a bunch of utility functions here that may be useful at least to reference: https://github.com/kubernetes/utils/blob/master/net/net.go

mcwumbly replied:
OK, updated this logic to match the table above with tests that should make it easier to review the various categories of cases here. Let me know if you have any additional thoughts on this auto-detection now with these changes.

fabriziopandini replied:
@mcwumby looks great to me! might be we should add a "configuring network" page in CAPI book under tasks, but this could be also in a separated PR.

@vincepri @CecileRobertMichon given that we are now identifying a set of invalid network configurations, should we error out in the validation web hook for the Cluster object?

mcwumbly replied:
FWIW, some discussion about this on sig-networking: https://groups.google.com/g/kubernetes-sig-network/c/S_64nw4BqoM/m/w64pw2ejAwAJ

mcwumbly commented:
Thanks for all the comments @fabriziopandini! we'll plan to follow up with another revision shortly -- I think we have a way forward with the auto-detection for now, but will continue to discuss if there are other ideas in the meantime.

yastij commented:
/cc

fabriziopandini commented:
/ok-to-test

fabriziopandini commented:
@mcwumbly I have made some thinking about how to setup E2E tests for ipv6, and assuming that for starting we are going to implement only a prepull job running ipv6 quick start for each PR.

In order to do so:
- It is required to change https://github.com/kubernetes-sigs/cluster-api/blob/eef7bb9227377c31df1d20a60c4d8d778346fa68/test/e2e/e2e_suite_test.go#L184-L188 adding a new `ipFamily` parameter, and this should flow down to the kind configuration
- the `ipFamily` parameter should come from a new `IP_FAMILY`  variable in `docker.yaml`, defaulting to ipv4; also `DOCKER_SERVICE_CIDRS` and `DOCKER_POD_CIDRS` should rollback to ipv4 values
- a new `pull-cluster-api-e2e-ipv6-main` prow job will override the `IP_FAMILY`  variable as well as the `DOCKER_SERVICE_CIDRS` and `DOCKER_POD_CIDRS` variables with ipv6 values (*)
- the `IP_FAMILY` variable should also determine which template we are going to use for the quick start job at https://github.com/kubernetes-sigs/cluster-api/blob/eef7bb9227377c31df1d20a60c4d8d778346fa68/test/e2e/quick_start.go#L82
- we should create a `cluster-template-ipv6` template flavour in `https://github.com/kubernetes-sigs/cluster-api/tree/master/test/e2e/data/infrastructure-docker/v1alpha4`

(*) In order to do so this PR should merge first, however you can temporary make CI to work in ipv6 mode by exporting ipv6 value overrides in `scripts/ci-e2e.sh`; for local testing, you can do the same in `test/e2e/Makefile`

I know this is a lot to take in, feel free to reach out to me so I can provide some context and also eventually help in breaking out this work in separated PRs.

mcwumbly commented:
> I know this is a lot to take in, feel free to reach out to me so I can provide some context and also eventually help in breaking out this work in separated PRs.

Sounds good. I opened this initially as a draft with a spike so we could all see where we might need changes like what you're suggesting more easily. I plan to address these issues and will reach out if I think breaking some of it out into separate PRs makes sense. 

I hope to get some time to go through this all more tomorrow or early next week.

fabriziopandini commented on test/e2e/data/infrastructure-docker/v1alpha4/cluster-template-ipv6/cluster-with-kcp.yaml:
```diff
@@ -0,0 +1,22 @@
+---
```
What about calling this file kcp-ipv6.yaml (there is no cluster here)?

mcwumbly replied:
sure, no strong opinions from me on this. fwiw, i think there are a handful of other cases like this in case anyone wants o make a separate fixup to address for style/consistency (e.g. kcp-scale-in).

fabriziopandini commented on test/e2e/data/infrastructure-docker/v1alpha4/cluster-template-ipv6/md.yaml:
```diff
@@ -0,0 +1,16 @@
+---
```
what about calling this file md-ipv6.yaml for consistency?

fabriziopandini reviewed (commented):
Two nits, but overall I like how things are shaping out!

fabriziopandini commented:
@sbueringer if you have some spare cycles, PTAL

sbueringer commented:
> @sbueringer if you have some spare review cycles, PTAL

Yup, have some time now. I'll take a look.

sbueringer commented on test/infrastructure/docker/docker/loadbalancer.go:
```diff
@@ -136,15 +153,18 @@ func (s *LoadBalancer) UpdateConfiguration(ctx context.Context) error {
 		return "", errors.WithStack(err)
 	}
-	if lbip4 == "" {
+	if s.ipFamily == clusterv1.IPv6Family {
+		lbip = lbip6
+	}
+	if lbip == "" {
```
```suggestion
	lbIPv4, lbIPv6, err := s.container.IP(ctx)
	if err != nil {
		return "", errors.WithStack(err)
	}
	var lbIP string
	if s.ipFamily == clusterv1.IPv6Family {
		lbIP = lbIPv6
	} else {
		lbIP = lbIPv4
	}
	if lbIP == "" {
		// if there is a load balancer container with the same name exists but is stopped, it may not have IP address associated with it.
		return "", errors.Errorf("load balancer IP cannot be empty: container %s does not have an associated IP address", s.containerName())
	}
	return lbIP, nil
```
nit: I would prefer it this way. I guess the part I don't like is calling the ipV4 war initially lbIP and then overwriting it with the ipV6. I think it's also a bit more expressive as it's immediately clear that the first returned IP is the ipv4
 (GitHub didn't let me select all relevant lines)

sbueringer reviewed (commented):
Sorry mostly nitpicking. I hope it's helpful anyway :)

Very nice implementation !! Fits nicely into the existing implementation.

mcwumbly commented:
Thanks for the review @sbueringer - happy to just address the nits as you see fit. I think maintainers should have veto rights over these sort of things. Will update as you prefer, and if you change you or anyone else changes their mind in the future about things (that are non-breaking), please feel free to make any further changes as you see fit.

Will push an update with the changes requested by tomorrow sometime.

mcwumbly commented on api/v1alpha4/cluster_types.go:
```diff
@@ -215,6 +216,88 @@ func (c *Cluster) SetConditions(conditions Conditions) {
+		return DualStackFamily, nil
+	}
+
+	if podsIPFamily != servicesIPFamily {
```
yes, i can make it `else if` instead if that seems more readable.

sbueringer replied:
Both variants are fine for me. I just wasn't entirely sure what the intention was. 

So this means:
* Pod: DualStack, Service: DualStack or IPv4 or IPv6 => valid
* Pod: IPv4 or IPv6, Service: DualStack or the other one of IPv4 / IPv6 => invalid 
?

mcwumbly replied:
I think the unit tests capture this intention pretty well, and the either implementation passes those tests.

sbueringer replied:
Sorry should have read your comment above more carefully. sgtm

mcwumbly commented:
@sbueringer @fabriziopandini - I addressed the feedback from today in two additional commits to aid in review. If things look good, I'll plan to squash and force push.

sbueringer commented on test/e2e/e2e_suite_test.go:
```diff
@@ -185,6 +185,7 @@ func setupBootstrapCluster(config *clusterctl.E2EConfig, scheme *runtime.Scheme,
 			Name:               config.ManagementClusterName,
 			RequiresDockerSock: config.HasDockerProvider(),
 			Images:             config.Images,
+			IPFamily:           config.GetVariable("IP_FAMILY"),
```
@mcwumbly I think you missed that one

mcwumbly replied:
will do this too, thanks.

sbueringer commented on test/infrastructure/docker/examples/simple-cluster-ipv6.yaml:
```diff
@@ -0,0 +1,126 @@
+        bindPort: 6443
+      nodeRegistration:
+        kubeletExtraArgs:
+          eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
```
@mcwumbly Can you make sure you also add the cgroup-driver flag to all relevant YAMLs?
xref: https://github.com/kubernetes-sigs/cluster-api/pull/4614/files

mcwumbly replied:
sure thing.

sbueringer commented:
@mcwumbly Looks mostly good :). I answered one of your comments, I think the IP_FAMILY would be nice and we just made a change to our CAPD YAMLs to also set the cgroup-driver kubeletExtraArgs flag to `cgroupfs` (so we need this also here on this PR)

fabriziopandini commented:
pending addressing last comment and squash commits, I'm ready to lgtm this PR
great work!

mcwumbly commented:
cool, I'll make one more pass to address the outstanding feedback on Monday. have a great weekend!

mcwumbly commented:
@sbueringer and @fabriziopandini  - I believe I have addressed all the remaining feedback. Branch has been rebased on master and squashed to a single commit.

sbueringer commented on test/e2e/config/docker.yaml:
```diff
@@ -115,8 +116,8 @@ variables:
   KUBERNETES_VERSION_UPGRADE_TO: "v1.19.1"
   KUBERNETES_VERSION_UPGRADE_FROM: "v1.18.2"
   DOCKER_SERVICE_DOMAIN: "cluster.local"
+  IP_FAMILY: "ipv4"
```
```suggestion
  IP_FAMILY: "IPv4"
```

sbueringer commented on test/framework/bootstrap/kind_util.go:
```diff
@@ -55,6 +58,10 @@ func CreateKindBootstrapClusterAndLoadImages(ctx context.Context, input CreateKi
 	if input.RequiresDockerSock {
 		options = append(options, WithDockerSockMount())
 	}
+	if input.IPFamily == "ipv6" {
```
```suggestion
	if input.IPFamily == "IPv6" {
```

sbueringer commented on test/e2e/quick_start.go:
```diff
@@ -72,14 +72,19 @@ func QuickStartSpec(ctx context.Context, inputGetter func() QuickStartSpecInput)
 		By("Creating a workload cluster")
 
+		flavor := clusterctl.DefaultFlavor
+		if input.E2EConfig.GetVariable(IPFamily) == "ipv6" {
+			flavor = "ipv6"
```
```suggestion
		if input.E2EConfig.GetVariable(IPFamily) == "IPv6" {
			flavor = "ipv6"
```

sbueringer reviewed (commented):
@mcwumbly last round of nits from me

sbueringer commented:
Thx, great work!
/lgtm

fabriziopandini commented:
/lgtm
/approve

yay! 🥳
